tests/Query/Filter/ScheduledDateField.test.ts

```diff
@@ -45,3 +53,25 @@ describe('sorting by scheduled', () => {
         expectTaskComparesAfter(new ScheduledDateField().createReverseSorter(), date1, date2);
     });
 });
+
+describe('scheduled date', () => {
+    it('in current week/month/year', () => {
+        currentPeriodsTestArray.forEach((p) => {
+            const filter = new ScheduledDateField().createFilterOrErrorMessage('scheduled in current ' + p);
+            testFilter(filter, new TaskBuilder().scheduledDate(null), false);
+            testFilter(filter, new TaskBuilder().scheduledDate(moment().format('YYYY-MM-DD')), true);
+            testFilter(filter, new TaskBuilder().scheduledDate(moment().startOf(p).format('YYYY-MM-DD')), true);
+            testFilter(filter, new TaskBuilder().scheduledDate(moment().endOf(p).format('YYYY-MM-DD')), true);
+            testFilter(
+                filter,
+                new TaskBuilder().scheduledDate(moment().startOf(p).subtract(1, 'second').format('YYYY-MM-DD')),
+                false,
+            );
+            testFilter(
+                filter,
+                new TaskBuilder().scheduledDate(moment().endOf(p).add(1, 'second').format('YYYY-MM-DD')),
+                false,
+            );
+        });
+    });
+});
```

Comment:
There's one aspect of writing tests that I think we haven't talked about yet: helping our future selves when a test we have written fails.

Let me explain.

### The problem

I see on GitHub that the currently-pushed code has one or more failing tests.

This is what shows in the [GitHub Action log](https://github.com/obsidian-tasks-group/obsidian-tasks/pull/1638/checks
):

![image](https://user-images.githubusercontent.com/4840096/217181320-f55a8b1e-534e-49ac-b0c6-79a598e61bb3.png)

There is nothing in that output that tells us which of week/month/year is the one that is failing.

And also we have no idea what the date values being used are.

We can look at the file names and line numbers in the error description, but we cannot tell which instance of the loop is the problem, and what data is being used.

### The solution

Here is a different way of using loops in our tests, which includes information about what is failing:

```ts
    it.each([
        ['low', Priority.Low, false],
        ['low', Priority.None, true],
        ['none', Priority.None, false],
        ['none', Priority.Medium, true],
        ['medium', Priority.None, true],
        ['medium', Priority.Medium, false],
        ['high', Priority.Medium, true],
        ['high', Priority.High, false],
    ])('priority is not %s (with %s)', (filter: string, input: Priority, expected: boolean) => {
        // TODO Use name of input priority instead of
        testTaskFilterForTaskWithPriority(`priority is not ${filter}`, input, expected);
    });
```

I don't have time to explain it fully, but there will be plenty of examples using `it.each` in this project and online...

In the above, the `%s` strings mean that these are the names of the tests generated:

![image](https://user-images.githubusercontent.com/4840096/217182646-5705a233-cd06-4d53-aa91-3bb1e8a2a54c.png)

So where are you doing doing `currentPeriodsTestArray.forEach`, please could you change all of them to use `it.each` and include information in the generated test names to supply info so that test failures supply enough info.

### How to avoid this

When we pair, you will experience writing the failing test first before implementing the code - and one of the many benefits of this is that, for every test you write, you see the error message if it fails.... And can review whether there would be enough info in the failure message to identify the cause if the test fails in future

### Possible explanation of the failure

I looked at the raw logs and see another failure:

![image](https://user-images.githubusercontent.com/4840096/217183437-3499f576-3535-442c-b95d-c16368251de2.png)

I think this may mean that because the test are running on American machines, the isoWeek is starting on Sunday.

This means that somehow we will need to work out how to control the start-of-week date in the test code. Perhaps setting the time zone or the locale.

I don't know how, but I do know that once we can do that, we can write tests both for weeks starting on Sunday and Monday, and these will be super valuable tests.

### Final thought...

We only discovered this because the tests have specific values. Yay for writing specific values in to tests!

Reply:
Thanks a lot for highlighting this. I had these thought in my head while going on with the development (especially the lack of clarity of a test result, btw, regardless being positive or negative) and you made it much clearer.

Reply:
Cycling, this morning, it occurred to me that on the future the low level tests could me made more informative to that if the expect-filter-type test failed, it could include the filter and the explanation in its error message. 

That would be a bonus extra, independent from the tests you are kindly currently writing.